File: parent/287 -- Find the Duplicate Number/mine2.go
package main

func findDuplicate(nums []int) int {
	slow, fast := 0, 0

	for {
		slow, fast = nums[slow], nums[nums[fast]]
		if slow == fast {
			break
		}
	}

	fast = 0
	for slow != fast {
		slow, fast = nums[slow], nums[fast]
	}

	return fast
}

File: parent/287 -- Find the Duplicate Number/mine2.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        # 2 pointer. Floyd's Cycle Detection Algo
        slow = fast = 0

        while True:
            slow, fast = nums[slow], nums[nums[fast]]
            if slow == fast:
                break
        fast = 0
        while slow != fast:
            slow, fast = nums[slow], nums[fast]
        
        return fast